polkadot/xcm/xcm-executor/src/lib.rs

```diff
@@ -945,6 +1070,9 @@ impl<Config: config::Config> XcmExecutor<Config> {
 				// should be executed.
 				let Some(weight) = Option::<Weight>::from(weight_limit) else { return Ok(()) };
 				let old_holding = self.holding.clone();
+				// Store the asset being used for fees, so delivery fees can access it.
+				self.asset_for_fees = Some(fees.id.clone());
```

Comment:
That's my main concern, and the reason why I did the "looping through all assets in holding" trick, which I know is bad.
If local executions had a `BuyExecution` instruction then this would be great.
My current thought is to create a new entrypoint `prepare_and_execute_with_fee_asset`.